### ISSUE TITLE
datadog: can not enable the processAgent without also enabling the orchestratorExplorer due to issue in the should-enable-k8s-resource-monitoring named template

### ISSUE THREAD

**Issue by michaelajr** [DataDog/helm-charts#292]
**Describe what happened:**

I can not enable the `processAgent` when the `orchestratorExplorer` is disabled. When ONLY the `processAgent` is enabled, you will see the following error in `processAgent` container log:

```
Cannot retrieve ConfigMap datadog/datadog-cluster-id: configmaps "datadog-cluster-id" is forbidden: User "system:serviceaccount:datadog:datadog-cluster-agent" cannot get resource "configmaps" in API group "" in the namespace "datadog"
Failed to generate or retrieve the cluster ID: configmaps "datadog-cluster-id" is forbidden: User "system:serviceaccount:datadog:datadog-cluster-agent" cannot get resource "configmaps" in API group "" in the namespace "datadog"
```

This happens because the named template that enables the RBAC required for the `processAgent`, requires the `orchestratorExplorer` to be enabled. 

https://github.com/DataDog/helm-charts/blob/7fb6a6ea7e079778d403ae284f3f1f45eaf3b1ac/charts/datadog/templates/_helpers.tpl#L351-L357

The `orchestratorExplorer` was `disabled` in previous versions. And I have to disable it due to DataDog/datadog-agent#8505 where the orchestration forwarder is not honoring proxy settings.

**Describe what you expected:**

As in previous chart versions, I would like to enable the `processAgent` without enabling the `orchestratorExplorer`. At least until DataDog/datadog-agent#8505 is fixed.

**Steps to reproduce the issue:**

Set `datadog.orchestratorExplorer.enabled` to `false`, and  `datadog.processAgent.enabled` to `true`.

```yaml
# values.yaml
datadog:
  orchestratorExplorer:
    enabled: false
  processAgent:
    enabled: true
```

Check the agent pod's `processAgent` container logs.

**Additional environment details (Operating System, Cloud provider, etc):**

AWS

**Comment by xlucas** [comment]
Hello! Thanks for reporting this.
So far I'm not able to reproduce this issue, so I'll need more details to investigate what could be happening, namely:
- the agent version
- the cluster agent version
- the helm chart version
- the content of `datadog/templates/daemonset.yaml` when running `helm template` (same args as `helm install`).

As far as I see this, `should-enable-k8s-resource-monitoring` would be used only if the orchestrator explorer and the cluster-agent are both enabled in helm values. If the orchestrator explorer feature is disabled then the code in the process-agent that access the configmap here should not run. So I wonder if the `DD_ORCHESTRATOR_EXPLORER_ENABLED` environment variable is somehow still set to `true` on the process-agent container and if so why. To understand a bit better, the previously mentioned helm template output will help spot this.

**Comment by michaelajr** [comment]
Hey @xlucas - thanks for the response. The info you requested is pasted below. But to clarify, I WANT to run the `process-agent`, but I CAN NOT run the `orchestrator explorer`. So I have disabled the `orchestrator explorer` and enabled the `process-agent`. In this configuration, the result of `should-enable-k8s-resource-monitoring` is `false` and therefor the RBAC that is required for the `process-agent` is never set up in this file:

https://github.com/DataDog/helm-charts/blob/240fb5f8bc443ac70652b15046556e4d73f0da50/charts/datadog/templates/cluster-agent-rbac.yaml#L101-L139

This will result in the error in OP.

Here is the info you requested:

Agent version:  7.29.0
Cluster-agent version: 1.13.0

```
# datadog/templates/daemonset.yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    app.kubernetes.io/component: agent
    app.kubernetes.io/instance: datadog
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: datadog
    app.kubernetes.io/version: '7'
    argocd.argoproj.io/instance: datadog
    helm.sh/chart: datadog-2.16.5
  name: datadog
  namespace: datadog
spec:
  selector:
    matchLabels:
      app: datadog
  template:
    metadata:
      annotations:
        checksum/autoconf-config: 74234e98afe7498fb5daf1f36ac2d78acc339464f950703b8c019892f982b90b
        checksum/checksd-config: 44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a
        checksum/clusteragent_token: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b
        checksum/confd-config: 19565fd9e0ac8cb368be01d3ce419c654b6818943d0b4f6070e187a25a791094
        checksum/install_info: 9194afc161702bc0fa9a2c62039fd5c513d580b37c78f3ae4ff7a8f149cb9515
      labels:
        app: datadog
        app.kubernetes.io/component: agent
        app.kubernetes.io/instance: datadog
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: datadog
      name: datadog
    spec:
      affinity: {}
      containers:
        - command:
            - agent
            - run
          env:
            - name: GODEBUG
              value: x509ignoreCN=0
            - name: DD_API_KEY
              valueFrom:
                secretKeyRef:
                  key: api-key
                  name: datadog-api-key
            - name: DD_KUBERNETES_KUBELET_HOST
              valueFrom:
                fieldRef:
                  fieldPath: status.hostIP
            - name: DD_CLUSTER_NAME
              value: redacted
            - name: DD_TAGS
              value: redacted
            - name: DD_KUBERNETES_NODE_LABELS_AS_TAGS
              value: >-
                {"beta.kubernetes.io/instance-type":"aws_instance_type","kubernetes.io/role":"kube_role"}
            - name: DD_KUBERNETES_POD_LABELS_AS_TAGS
              value: redacted
            - name: DD_KUBERNETES_POD_ANNOTATIONS_AS_TAGS
              value: redacted
            - name: KUBERNETES
              value: 'yes'
            - name: DD_LOG_LEVEL
              value: INFO
            - name: DD_DOGSTATSD_PORT
              value: '8125'
            - name: DD_DOGSTATSD_NON_LOCAL_TRAFFIC
              value: 'true'
            - name: DD_DOGSTATSD_ORIGIN_DETECTION
              value: 'true'
            - name: DD_DOGSTATSD_TAG_CARDINALITY
              value: low
            - name: DD_CLUSTER_AGENT_ENABLED
              value: 'true'
            - name: DD_CLUSTER_AGENT_KUBERNETES_SERVICE_NAME
              value: datadog-cluster-agent
            - name: DD_CLUSTER_AGENT_AUTH_TOKEN
              valueFrom:
                secretKeyRef:
                  key: token
                  name: datadog-cluster-agent-token
            - name: DD_APM_ENABLED
              value: 'false'
            - name: DD_LOGS_ENABLED
              value: 'false'
            - name: DD_LOGS_CONFIG_CONTAINER_COLLECT_ALL
              value: 'false'
            - name: DD_LOGS_CONFIG_K8S_CONTAINER_USE_FILE
              value: 'true'
            - name: DD_HEALTH_PORT
              value: '5555'
            - name: DD_DOGSTATSD_SOCKET
              value: /var/run/datadog/dsd.socket
            - name: DD_EXTRA_CONFIG_PROVIDERS
              value: clusterchecks endpointschecks
            - name: DD_IGNORE_AUTOCONF
              value: kubernetes_state
            - name: HTTP_PROXY
              value: redacted
            - name: HTTPS_PROXY
              value: redacted
            - name: NO_PROXY
              value: redacted
            - name: DD_KUBELET_TLS_VERIFY
              value: 'false'
          image: 'redatced/datadog/agent:7.29.0'
          imagePullPolicy: IfNotPresent
          livenessProbe:
            failureThreshold: 6
            httpGet:
              path: /live
              port: 5555
              scheme: HTTP
            initialDelaySeconds: 15
            periodSeconds: 15
            successThreshold: 1
            timeoutSeconds: 5
          name: agent
          ports:
            - containerPort: 8125
              name: dogstatsdport
              protocol: UDP
          readinessProbe:
            failureThreshold: 6
            httpGet:
              path: /ready
              port: 5555
              scheme: HTTP
            initialDelaySeconds: 15
            periodSeconds: 15
            successThreshold: 1
            timeoutSeconds: 5
          resources:
            limits:
              cpu: 500m
              memory: 768Mi
            requests:
              cpu: 200m
              memory: 256Mi
          volumeMounts:
            - mountPath: /etc/datadog-agent/install_info
              name: installinfo
              readOnly: true
              subPath: install_info
            - mountPath: /var/log/datadog
              name: logdatadog
            - mountPath: /tmp
              name: tmpdir
              readOnly: false
            - mountPath: /etc/datadog-agent
              name: config
            - mountPath: /host/var/run
              mountPropagation: None
              name: runtimesocketdir
              readOnly: true
            - mountPath: /var/run/datadog
              name: dsdsocket
            - mountPath: /host/proc
              mountPropagation: None
              name: procdir
              readOnly: true
            - mountPath: /host/sys/fs/cgroup
              mountPropagation: None
              name: cgroups
              readOnly: true
            - mountPath: /etc/datadog-agent/conf.d/kube_apiserver_metrics.d
              name: kube-apiserver-metrics
        - command:
            - process-agent
            - '-config=/etc/datadog-agent/datadog.yaml'
          env:
            - name: GODEBUG
              value: x509ignoreCN=0
            - name: DD_API_KEY
              valueFrom:
                secretKeyRef:
                  key: api-key
                  name: datadog-api-key
            - name: DD_KUBERNETES_KUBELET_HOST
              valueFrom:
                fieldRef:
                  fieldPath: status.hostIP
            - name: DD_CLUSTER_NAME
              value: redacted
            - name: DD_TAGS
              value: redacted
            - name: DD_KUBERNETES_NODE_LABELS_AS_TAGS
              value: redacted
            - name: DD_KUBERNETES_POD_LABELS_AS_TAGS
              value: redacted
            - name: DD_KUBERNETES_POD_ANNOTATIONS_AS_TAGS
              value: '{"iam.amazonaws.com/role":"kube_iamrole"}'
            - name: KUBERNETES
              value: 'yes'
            - name: DD_CLUSTER_AGENT_ENABLED
              value: 'true'
            - name: DD_CLUSTER_AGENT_KUBERNETES_SERVICE_NAME
              value: datadog-cluster-agent
            - name: DD_CLUSTER_AGENT_AUTH_TOKEN
              valueFrom:
                secretKeyRef:
                  key: token
                  name: datadog-cluster-agent-token
            - name: DD_LOG_LEVEL
              value: INFO
            - name: DD_SYSTEM_PROBE_ENABLED
              value: 'false'
            - name: DD_ORCHESTRATOR_EXPLORER_ENABLED
              value: 'false'
            - name: HTTP_PROXY
              value: redacted
            - name: HTTPS_PROXY
              value: redacted
            - name: NO_PROXY
              value: redacted
            - name: DD_KUBELET_TLS_VERIFY
              value: 'false'
          image: 'redacted/datadog/agent:7.29.0'
          imagePullPolicy: IfNotPresent
          name: process-agent
          resources:
            limits:
              cpu: 200m
              memory: 512Mi
            requests:
              cpu: 100m
              memory: 256Mi
          volumeMounts:
            - mountPath: /etc/datadog-agent
              name: config
            - mountPath: /host/var/run
              mountPropagation: None
              name: runtimesocketdir
              readOnly: true
            - mountPath: /var/log/datadog
              name: logdatadog
            - mountPath: /tmp
              name: tmpdir
              readOnly: false
            - mountPath: /host/sys/fs/cgroup
              mountPropagation: None
              name: cgroups
              readOnly: true
            - mountPath: /etc/passwd
              name: passwd
              readOnly: true
            - mountPath: /host/proc
              mountPropagation: None
              name: procdir
              readOnly: true
            - mountPath: /etc/datadog-agent/conf.d/kube_apiserver_metrics.d
              name: kube-apiserver-metrics
      hostPID: true
      initContainers:
        - args:
            - cp -r /etc/datadog-agent /opt
          command:
            - bash
            - '-c'
          image: 'redacted/agent:7.29.0'
          imagePullPolicy: IfNotPresent
          name: init-volume
          resources: {}
          volumeMounts:
            - mountPath: /opt/datadog-agent
              name: config
        - args:
            - >-
              for script in $(find /etc/cont-init.d/ -type f -name '*.sh' |
              sort) ; do bash $script ; done
          command:
            - bash
            - '-c'
          env:
            - name: GODEBUG
              value: x509ignoreCN=0
            - name: DD_API_KEY
              valueFrom:
                secretKeyRef:
                  key: api-key
                  name: datadog-api-key
            - name: DD_KUBERNETES_KUBELET_HOST
              valueFrom:
                fieldRef:
                  fieldPath: status.hostIP
            - name: DD_CLUSTER_NAME
              value: redacted
            - name: DD_TAGS
              value: redacted
            - name: DD_KUBERNETES_NODE_LABELS_AS_TAGS
              value: redacted
            - name: DD_KUBERNETES_POD_LABELS_AS_TAGS
              value: redacted
            - name: DD_KUBERNETES_POD_ANNOTATIONS_AS_TAGS
              value: '{"iam.amazonaws.com/role":"kube_iamrole"}'
            - name: KUBERNETES
              value: 'yes'
          image: 'redacted/datadog/agent:7.29.0'
          imagePullPolicy: IfNotPresent
          name: init-config
          resources: {}
          volumeMounts:
            - mountPath: /var/log/datadog
              name: logdatadog
            - mountPath: /etc/datadog-agent
              name: config
            - mountPath: /conf.d
              name: confd
              readOnly: true
            - mountPath: /host/proc
              mountPropagation: None
              name: procdir
              readOnly: true
            - mountPath: /host/var/run
              mountPropagation: None
              name: runtimesocketdir
              readOnly: true
      nodeSelector:
        kubernetes.io/os: linux
      priorityClassName: redacted
      serviceAccountName: datadog
      tolerations:
        - effect: NoSchedule
          key: node-role.kubernetes.io/master
      volumes:
        - configMap:
            name: datadog-installinfo
          name: installinfo
        - emptyDir: {}
          name: config
        - hostPath:
            path: /var/run
          name: runtimesocketdir
        - emptyDir: {}
          name: logdatadog
        - emptyDir: {}
          name: tmpdir
        - hostPath:
            path: /proc
          name: procdir
        - hostPath:
            path: /sys/fs/cgroup
          name: cgroups
        - hostPath:
            path: /var/run/datadog/
            type: DirectoryOrCreate
          name: dsdsocket
        - emptyDir: {}
          name: s6-run
        - configMap:
            name: datadog-confd
          name: confd
        - hostPath:
            path: /etc/passwd
          name: passwd
        - emptyDir: {}
          name: kube-apiserver-metrics
  updateStrategy:
    rollingUpdate:
      maxUnavailable: 50%
    type: RollingUpdate
```

```
```

**Comment by michaelajr** [comment]
I should also note that removing the conditional in `templates/cluster-agent-rbac.yaml` at line `101` solved my issue.

**Comment by xlucas** [comment]
Thanks for the detail! I've been digging into this and been able to reproduce the issue I believe. What happens here is, once the orchestratorExplorer feature is disabled:
- the cluster-agent doesn't have the RBAC set to access the `datadog-cluster-id` configmap that stores the clusterID used to uniquely identify collected k8s resources
- the process-agent makes a call to the cluster-agent in order to retrieve the clusterID when the feature is enabled, because it is responsible for collecting pods (the cluster-agent collects all other resources). A code path was overlooked in that agent and is still making the call even though the feature is disabled. It ends up with an error each time because the cluster-agent hasn't the appropriate RBAC set in that unexpected scenario.

Although this will log a good amount of errors on the cluster-agent and be noisy over time, this is not preventing it from functioning. This is also not preventing the process-agent from functioning, because it just skips those errors.

The PR to fix this issue on the process-agent side is the following: https://github.com/DataDog/datadog-agent/pull/8544

**Comment by xlucas** [comment]
closing this as the previously mentionned PR was included in Agent 7.30.0 + another fix https://github.com/DataDog/helm-charts/pull/321 made it to the helm chart which should have resolved the issue now

**Comment by michaelajr** [comment]
Thanks @xlucas. Sorry I never got back to you on the fix. We patched our local copy. Will upgrade to the latest chart. Thanks again.
